src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +697,232 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
+
+			IndexedIO::EntryIDList tags;
+			ConstIndexedIOPtr tagsIO = m_indexedIO->subdirectory( localTagsEntry, IndexedIO::NullIfMissing );
+			if( tagsIO )
+			{
+				tagsIO->entryIds( tags );
+			}
+
+			IndexedIO::EntryIDList descendentTags;
+			ConstIndexedIOPtr descendentTagsIO = m_indexedIO->subdirectory( descendentTagsEntry, IndexedIO::NullIfMissing );
+			if( descendentTagsIO )
+			{
+				descendentTagsIO->entryIds( descendentTags );
+			}
+
+			/// provided for backward compatibility...
+			if( tags.empty() && descendentTags.empty() )
+			{
+				ConstIndexedIOPtr tagsIO = m_indexedIO->subdirectory( tagsEntry, IndexedIO::NullIfMissing );
+				if( tagsIO )
+				{
+					tagsIO->entryIds( tags, IndexedIO::File );
+					tagsIO->entryIds( descendentTags );
+				}
+			}
+
+			if( tags.empty() && descendentTags.empty() )
+			{
+				return;
+			}
+
+			m_sets = new CompoundData();
+
+			SceneInterface::Path currentPath;
+			path( currentPath );
+			for( const auto &tag : tags )
+			{
+				getRoot()->appendToSet( tag, currentPath );
+			}
+
+			if( !descendentTags.empty() )
+			{
+				SceneCache::NameList children;
+				childNames( children );
+
+				for( const auto &c : children )
+				{
+					ReaderImplementationPtr childReader = child( c, NullIfMissing );
+					if( childReader )
+					{
+						childReader->convertTagsToSets();
+					}
+				}
+			}
+		}
+
 	private :
 
+		void appendToSet( const Name &setName, const SceneInterface::Path &path )
+		{
+			auto &writable = m_sets->writable();
+			auto it = writable.find( setName );
+			if( it == writable.end() )
+			{
+				it = writable.insert( std::make_pair( setName, new IECore::PathMatcherData() ) ).first;
+			}
+			PathMatcherDataPtr pathMatcher = runTimeCast<PathMatcherData>( it->second );
+			pathMatcher->writable().addPath( path );
+		}
+
+		PathMatcherDataPtr readLocalSet( const Name &name ) const
+		{
+			IECore::CompoundDataBasePtr compoundDataBase = getSetCompound();
+
+			if( !compoundDataBase )
+			{
+				return new IECore::PathMatcherData();
+			}
+
+			CompoundDataPtr compoundData = new CompoundData( compoundDataBase->readable() );
+
+			if( IECore::PathMatcherDataPtr foundPathMatcher = compoundData->member<PathMatcherData>( name ) )
+			{
+				return foundPathMatcher;
+			}
+
+			return new IECore::PathMatcherData();
+		}
+
+		NameList readChildSetNames() const
+		{
+			IndexedIO::EntryIDList childSetNames;
+			IndexedIOPtr childSets = m_indexedIO->subdirectory( childSetsEntry, IECore::IndexedIO::NullIfMissing );
+			if( childSets )
+			{
+				childSets->entryIds( childSetNames );
+			}
+			else
+			{
+				IndexedIOPtr descenentTags = m_indexedIO->subdirectory( descendentTagsEntry, IECore::IndexedIO::NullIfMissing);
+				if ( descenentTags )
+				{
+					descenentTags->entryIds( childSetNames );
+				}
+			}
+
+			return childSetNames;
+		}
+
+		IECore::CompoundDataBasePtr getSetCompound() const
+		{
+			if( m_sets )
+			{
+				return runTimeCast<CompoundDataBase>( m_sets );
```

Comment:
This looks like the intention is to avoid the work done later in this function, but then `m_sets` isn't updated later in this function at all. What _is_ the intention?

Reply:
Looks like you missed this comment? For me, this was the crux for understanding (or not) what was going on here. I really couldn't follow what this was used for and when. My best guess was that it was only for converted tags, and that it would only ever be updated on the root?

I realise I actually don't even know why we store sets at all. It doesn't seem too unreasonable to just pay the cost whenever they are loaded, especially from a Gaffer standpoint where they'll be externally Gaffer anyway. If we were to cache them, then we should probably be doing it in a similar vein to `SharedData::objectCache` and `SharedData::transformCache`. These use a ComputationCache/ObjectPool combo for storage, so we have external control over how much memory is used, and we use an LRU strategy for throwing away stuff that hasn't been used for a while.

I think it'd be great to start with a much simpler stateless (as in no m_sets) implementation that just loads on demand and is easy to verify.